nano/node/repcrawler.cpp

```diff
@@ -1,370 +1,524 @@
+		bool const sufficient_weight = current_total_weight > node.online_reps.delta ();
+
+		// If online weight drops below minimum, reach out to preconfigured peers
+		if (!sufficient_weight)
+		{
+			stats.inc (nano::stat::type::rep_crawler, nano::stat::detail::keepalive);
+			node.keepalive_preconfigured ();
+		}
+
+		lock.lock ();
+
+		condition.wait_for (lock, query_interval (sufficient_weight), [this, sufficient_weight] {
+			return stopped || query_predicate (sufficient_weight) || !responses.empty ();
+		});
+
+		if (stopped)
+		{
+			return;
+		}
+
+		stats.inc (nano::stat::type::rep_crawler, nano::stat::detail::loop);
+
+		if (!responses.empty ())
+		{
+			validate_and_process (lock);
+			debug_assert (!lock.owns_lock ());
+			lock.lock ();
+		}
+
+		cleanup ();
+
+		if (query_predicate (sufficient_weight))
+		{
+			last_query = std::chrono::steady_clock::now ();
+
+			auto targets = prepare_crawl_targets (sufficient_weight);
+
+			lock.unlock ();
+			query (targets);
+			lock.lock ();
+		}
+
+		debug_assert (lock.owns_lock ());
 	}
-	// Reduce crawl frequency when there's enough total peer weight
-	unsigned next_run_ms = node.network_params.network.is_dev_network () ? 100 : sufficient_weight ? 7000
-																								   : 3000;
-	std::weak_ptr<nano::node> node_w (node.shared ());
-	node.workers.add_timed_task (now + std::chrono::milliseconds (next_run_ms), [node_w, this] () {
-		if (auto node_l = node_w.lock ())
+}
+
+void nano::rep_crawler::cleanup ()
+{
+	debug_assert (!mutex.try_lock ());
+
+	// Evict reps with dead channels
+	erase_if (reps, [this] (rep_entry const & rep) {
+		if (!rep.channel->alive ())
+		{
+			logger.info (nano::log::type::rep_crawler, "Evicting representative {} with dead channel at {}", rep.account.to_account (), rep.channel->to_string ());
+			stats.inc (nano::stat::type::rep_crawler, nano::stat::detail::channel_dead);
+			return true; // Erase
+		}
+		return false;
+	});
+
+	// Evict queries that haven't been responded to in a while
+	erase_if (queries, [this] (query_entry const & query) {
+		if (nano::elapsed (query.time, config.query_timeout))
 		{
-			this->ongoing_crawl ();
+			logger.debug (nano::log::type::rep_crawler, "Aborting unresponsive query for block {} from {}", query.hash.to_string (), query.channel->to_string ());
+			stats.inc (nano::stat::type::rep_crawler, nano::stat::detail::query_timeout);
+			return true; // Erase
 		}
+		return false;
 	});
 }
 
-std::vector<std::shared_ptr<nano::transport::channel>> nano::rep_crawler::get_crawl_targets (nano::uint128_t total_weight_a)
+std::vector<std::shared_ptr<nano::transport::channel>> nano::rep_crawler::prepare_crawl_targets (bool sufficient_weight) const
 {
-	constexpr std::size_t conservative_count = 10;
-	constexpr std::size_t aggressive_count = 40;
+	debug_assert (!mutex.try_lock ());
+
+	// TODO: Make these values configurable
+	constexpr std::size_t conservative_count = 160;
+	constexpr std::size_t aggressive_count = 160;
```

Comment:
The conservative count is the same as the aggressive count. Is this a mistake?
Also, these numbers are much bigger than the previous change.